contractsV5/eth2dai/Eth2DaiReserve.sol

```diff
@@ -0,0 +1,846 @@
+    // basicData contains compact data of min eth support, max traverse and max takes
+    // min eth support (first 192 bits) + max traverse (32 bits) + max takes (32 bits) = 256 bits
+    mapping(address => uint) internal tokenBasicData;
+    // factorData contains compact data of factors to compute max traverse, max takes, and min take order size
+    // 6 params, each 32 bits (6 * 32 = 192 bits)
+    mapping(address => uint) internal tokenFactorData;
+
+    struct BasicDataConfig {
+        uint minETHSupport;
+        uint maxTraverse;
+        uint maxTakes;
+    }
+
+    struct FactorDataConfig {
+        uint maxTraverseX;
+        uint maxTraverseY;
+        uint maxTakeX;
+        uint maxTakeY;
+        uint minOrderSizeX;
+        uint minOrderSizeY;
+    }
+
+    struct InternalInventoryData {
+        uint minTokenBal;
+        uint maxTokenBal;
+        uint premiumBps;
+        uint minSpreadBps;
+    }
+
+    struct OfferData {
+        uint payAmount;
+        uint buyAmount;
+        uint id;
+    }
+
+    constructor(address _kyberNetwork, uint _feeBps, address _otc, address _weth, address _admin) public {
+        require(_kyberNetwork != address(0), "constructor: kyberNetwork's address is missing");
+        require(_otc != address(0), "constructor: otc's address is missing");
+        require(_weth != address(0), "constructor: weth's address is missing");
+        require(_feeBps < 10000, "constructor: fee >= 10000");
+        require(_admin != address(0), "constructor: admin is missing");
+
+        wethToken = WethInterface(_weth);
+        require(getDecimals(wethToken) == MAX_DECIMALS, "constructor: wethToken's decimals is not MAX_DECIMALS");
+        require(wethToken.approve(_otc, 2**255), "constructor: failed to approve otc (wethToken)");
+    
+        kyberNetwork = _kyberNetwork;
+        otc = OtcInterface(_otc);
+        feeBps = _feeBps;
+        admin = _admin;
+        tradeEnabled = true;
+    }
+
+    function() external payable {} // solhint-disable-line no-empty-blocks
+
+    /**
+        Returns conversion rate of given pair and srcQty, use 1 as srcQty if srcQty = 0
+        Using eth amount to compute offer limit configurations
+        => need to check spread is ok for token -> eth
+        Last bit of the rate indicates whether to use internal inventory:
+          0 - use eth2dai
+          1 - use internal inventory
+    */
+    function getConversionRate(ERC20 src, ERC20 dest, uint srcQty, uint) public view returns(uint) {
+        if (!tradeEnabled) { return 0; }
+        // check if token's listed
+        ERC20 token = src == ETH_TOKEN_ADDRESS ? dest : src;
+        if (!isTokenListed[address(token)]) { return 0; }
+        
+        OfferData memory bid;
+        OfferData memory ask;
+        (bid, ask) = getFirstBidAndAskOrders(token);
+
+        // if token is src, need to check for valid spread, 
+        if (token == src && !checkValidSpread(bid, ask, false, 0)) { return 0; }
```

Comment:
don't recall why we do this?

Reply:
We need to calculate equivalent eth amount for trade from token -> eth (for calculating offer limits, internal inventory check)
So we return 0 if the spread is not ok for swapping token -> eth